ops/testing.py

```diff
@@ -0,0 +1,408 @@
+        """
+        # TODO: jam 2020-03-05 We probably want to take config as a parameter as well, since
+        #       it would define the default values of config that the charm would see.
+        self._charm_cls = charm_cls
+        self._charm = None
+        self._meta = self._create_meta(meta)
+        self._unit_name = self._meta.name + '/0'
+        self._model = None
+        self._framework = None
+        self._hooks_enabled = True
+        self._relation_id_counter = 0
+        self._backend = _TestingModelBackend(self._unit_name)
+        self._model = model.Model(self._unit_name, self._meta, self._backend)
+        self._framework = framework.Framework(":memory:", "no-disk-path", self._meta, self._model)
+
+    @property
+    def charm(self):
+        return self._charm
+
+    @property
+    def model(self):
+        return self._model
+
+    @property
+    def framework(self):
+        return self._framework
+
+    def begin(self):
+        """Instantiate the Charm and start handling events.
+
+        Before calling begin(), changes to the model won't be
+        """
+        # The Framework adds attributes to class objects for events, etc. As such, we can't re-use
+        # the original class against multiple Frameworks. So create a locally defined class
+        # and register it.
+        # TODO: jam 2020-03-16 We are looking to changes this to Instance attributes instead of
+        #       Class attributes which should clean up this ugliness. The API can stay the same
+        class TestEvents(self._charm_cls.on.__class__):
+            pass
+
+        TestEvents.__name__ = self._charm_cls.on.__class__.__name__
+
+        class TestCharm(self._charm_cls):
+            on = TestEvents()
+
+        # Note: jam 2020-03-01 This is so that errors in testing say MyCharm has no attribute foo,
+        # rather than TestCharm has no attribute foo.
+        TestCharm.__name__ = self._charm_cls.__name__
+        self._charm = TestCharm(self._framework, self._framework.meta.name)
+
+    def _create_meta(self, charm_metadata):
+        """Create a CharmMeta object.
+
+        Handle the cases where a user doesn't supply an explicit metadata snippet.
+        """
+        if charm_metadata is None:
+            filename = inspect.getfile(self._charm_cls)
+            metadata_path = pathlib.Path(filename).parents[1] / 'metadata.yaml'
+            if metadata_path.is_file():
+                charm_metadata = metadata_path.read_text()
+            else:
+                # The simplest of metadata that the framework can support
+                charm_metadata = 'name: test-charm'
+        elif isinstance(charm_metadata, str):
+            charm_metadata = dedent(charm_metadata)
+        return charm.CharmMeta.from_yaml(charm_metadata)
+
+    def enable_hooks(self):
+        """Start emitting hook events from charm.on when the model is changed.
+
+        Once enable_hooks is passed the charm, any changes to the model (such as
```

Comment:
I think the "is passed the charm" bit is leftover from previous iterations.
Also I'd move `enable_hooks` after `disable_hooks` (so in the docs they're in the more natural order), and I'd mention something about enable_hooks undoing what disable_hooks did, as hooks are enabled by default once you've begun